packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx

```diff
@@ -189,61 +197,80 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
       const targetDay = itemProps.value;
       setSelectedDate(targetDay.originalDate);
-      setOpenState(OpenState.Closing);
+      setOpenState(false);
       setError('');
       setFormattedDate(valueFormatter(targetDay.originalDate));
 
-      _.invoke(props, 'onDateChange', e, { ...props, value: targetDay.originalDate });
+      _.invoke(predefinedProps, 'onDateChange', e, { itemProps, value: targetDay.originalDate });
     },
   });
 
   const calendarElement = createShorthand(DatepickerCalendar, calendar, {
-    defaultProps: () => getA11yProps('calendar', {}),
+    defaultProps: () => getA11yProps('calendar', { ...calendarOptions, ...dateFormatting }),
     overrideProps: overrideDatepickerCalendarProps,
   });
 
-  const openStateToBooleanKnob = (openState: OpenState): boolean => {
-    return openState === OpenState.Open;
-  };
-
-  const onInputClick = (): void => {
-    if (openState === OpenState.Closed) {
-      setOpenState(OpenState.Open);
-    } else if (openState === OpenState.Open || openState === OpenState.Closing) {
-      // Keep popup open in case we can only enter the date through calendar.
-      if (props.allowManualInput) {
-        setOpenState(OpenState.Closed);
-      } else {
-        setOpenState(OpenState.Open);
+  const overrideInputProps = (predefinedProps: InputProps): InputProps => ({
+    onClick: (e): void => {
+      if (preventOpeningOnClick && !openState) {
+        setPreventOpeningOnClick(false);
+      } else if (!openState) {
+        setOpenState(true);
+      } // Keep popup open in case we can only enter the date through calendar.
+      else if (props.allowManualInput) {
+        setOpenState(false);
       }
-    }
-  };
 
-  const onInputChange = (e, target: { value: string }) => {
-    const parsedDate = props.parseDate(target.value);
-
-    setFormattedDate(target.value);
-    if (parsedDate) {
-      if (isRestrictedDate(parsedDate, calendarOptions)) {
-        setError(props.isOutOfBoundsErrorMessage);
-      } else {
-        setError('');
+      _.invoke(predefinedProps, 'onClick', e, predefinedProps);
+    },
+    onChange: (e, target: { value: string }) => {
+      const parsedDate = props.parseDate(target.value);
+      const validationError = validateDate(parsedDate, target.value, calendarOptions, dateFormatting, props.required);
+      setError(validationError);
+      setFormattedDate(target.value);
+      if (!validationError && !!parsedDate) {
         setSelectedDate(parsedDate);
         _.invoke(props, 'onDateChange', e, { ...props, value: parsedDate });
       }
-    } else if (target.value) {
-      setError(props.invalidInputErrorMessage);
-    } else if (props.required && !selectedDate) {
-      setError(props.isRequiredErrorMessage);
-    } else {
-      setError('');
-    }
-  };
+
+      if (!!validationError) {
+        _.invoke(props, 'onDateChangeError', e, { ...props, error: validationError });
+      }
+
+      _.invoke(predefinedProps, 'onChange', e, target);
```

Comment:
Not sure if it's correct to trigger both onChange and onDateChangeError 
@layershifter WDYT?